emukit/core/bandit_parameter.py

```diff
@@ -0,0 +1,140 @@
+                # NOTE Categorical sub-parameters not yet implemented because inputs are
+                # homogeneously typed np.ndarrays rather than structured arrays. In the future,
+                # using structured arrays for all inputs may be more appropriate.
+            parameters.append(parameter)
+        return(parameters)
+
+    def check_in_domain(self, x: Union[np.ndarray, float, list]) -> Union[bool, np.ndarray]:
+        """
+        Checks if all the points in x lie in the domain set
+
+        :param x:    1d numpy array of points to check
+                  or 2d numpy array with shape (n_points, 1) of points to check
+                  or float of single point to check
+        :return: A boolean value which indicates whether all points lie in the domain
+        """
+        if isinstance(x, np.ndarray):
+            if x.ndim == 2 and x.shape[1] == 1:
+                x = x.ravel()
+            elif x.ndim == 2 and x.shape[1] > 1:
+                result = np.array([self.check_in_domain(xx) for xx in x])
+                return result
+            elif x.ndim > 1:
+                raise ValueError("Expected x shape (n,) or (n, 1), actual is {}".format(x.shape))
+            if x.shape[0] != self.domain.shape[1]:
+                raise ValueError("Received x with dimension {}, expected dimension is {}".format(x.shape[0],self.domain.shape[1]))
+            result = (self.domain == x).all(axis=1).any()
+        elif isinstance(x, float):
+            if self.domain.shape[1] > 1:
+                raise ValueError("Received x with dimension 1, expected dimension is {}".format(x.shape[0],self.domain.shape[1]))
+            result = (self.domain == x).all(axis=1).any()
+        elif isinstance(x, list):
+            result = (self.domain == x).all(axis=1).any()
+        else:
+            raise ValueError("Unsupported type for point x: {}".format(type(x)))
+        return result
+
+    @property
+    def bounds(self) -> List[Tuple]:
+        """
+        Returns a list containing the bounds for each constituent parameter
+        """
+        return [pb for p in self.parameters for pb in p.bounds]
+
+    def round(self, x: np.ndarray) -> np.ndarray:
+        """
+        Rounds each row in x to represent a valid value for this bandit variable. Note that this
+        valid value may be 'far' from the suggested value.
+
+        :param x: A 2d array NxD to be rounded (D is len(self.parameters))
+        :returns: An array NxD where each row represents a value from the domain
+                  that is closest to the corresponding row in x
+        """
+        if x.ndim != 2:
+            raise ValueError("Expected 2d array, got " + str(x.ndim))
+
+        if x.shape[1] != self.dimension:
+            raise ValueError("Expected {} column array, got {}".format(self.dimension, x.shape[1]))
+
+        x_rounded = []
+        for row in x:
+            dists = np.sqrt(np.sum((self.domain - row)**2))
```

Comment:
is numerical stability a concern here? (can the value under sqrt be small and negative)

Reply:
Unless the `self.domain` and `row` values are complex, those are sums of squares of real numbers. So no the summands will not be negative -> the sum will be non-negative.
Numerical stability could be a concern here, and I'm pretty sure there's `numpy.linalg.norm` which is a potentially better option.

Reply:
ah, of course, didn't notice that, sorry. 
I recall scipy.linalg.norm is more stable but I think it's fine as is.